pkg/gridtypes/zos/vm.go

```diff
@@ -0,0 +1,98 @@
+package zos
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/threefoldtech/zos/pkg/gridtypes"
+)
+
+// VirtualMachine reservation data
+type VirtualMachine struct {
+	Name     string   `json:"name"`
+	Network  string   `json:"network"`
+	IP       net.IP   `json:"ip"`
+	SSHKeys  []string `json:"ssh_keys"`
+	PublicIP string   `json:"public_ip"`
+	Size     uint8    `json:"size"`
+}
+
+// Valid implementation
+func (v VirtualMachine) Valid(getter gridtypes.WorkloadGetter) error {
+	if matched, _ := regexp.MatchString("^[0-9a-zA-Z-.]*$", v.Name); !matched {
+		return errors.New("the name must consist of alphanumeric characters, dot, and dash ony")
+	}
+
+	for _, key := range v.SSHKeys {
+		trimmed := strings.TrimSpace(key)
+		if strings.ContainsAny(trimmed, "\t\r\n\f\"") {
+			return errors.New("ssh keys can't contain intermediate whitespace chars or quotes other than white space")
+		}
+	}
+
+	if v.IP.To4() == nil && v.IP.To16() == nil {
+		return errors.New("invalid IP")
+	}
+	if v.Size < 1 || v.Size > 18 {
+		return errors.New("unsupported vm size %d, only size 1 to 18 are supported")
+	}
+	wl, err := getter.Get(v.PublicIP)
+	if err != nil {
+		return fmt.Errorf("public ip is not found")
+	}
+
+	if wl.Type != PublicIPType {
+		return errors.Wrapf(err, "workload of name '%s' is not a public ip", v.PublicIP)
+	}
+
+	return nil
+}
+
+// Capacity implementation
+func (v VirtualMachine) Capacity() (gridtypes.Capacity, error) {
```

Comment:
Since Capacity is exactly the same for K8s and Vm, the implementation from K8s should be deleted. the way Composition (inheritance) works in Go will make this method available for k8s

Reply:
It's only implemented in vm?